From b491ff5f679e5e8d6330af735fd73b36b0b9f7c0 Mon Sep 17 00:00:00 2001
From: Jeevaka Prabu Badrappan <jeevaka.badrappan@intel.com>
Date: Sun, 29 Mar 2020 15:44:15 +0530
Subject: [PATCH] BACKPORT: Policies to get Android booted with 5.4 & above
 kernels

Following patches are ported:
update sepolicy for fs notification hooks
neverallow_macros: add watch* perms
global_macros: trim back various watch* permissions
access_vectors: re-organize common file perms

Tracked-On:
Signed-off-by: Jeevaka Prabu Badrappan <jeevaka.badrappan@intel.com>
---
 private/access_vectors   | 38 +++++++++-----------------------------
 public/global_macros     |  4 ++--
 public/neverallow_macros |  2 +-
 3 files changed, 12 insertions(+), 32 deletions(-)

diff --git a/private/access_vectors b/private/access_vectors
index b77dcc1f7..7e5e69eda 100644
--- a/private/access_vectors
+++ b/private/access_vectors
@@ -27,6 +27,14 @@ common file
 	execute
 	quotaon
 	mounton
+	audit_access
+	open
+	execmod
+	watch
+	watch_mount
+	watch_sb
+	watch_with_perm
+	watch_reads
 }
 
 
@@ -153,6 +161,7 @@ class filesystem
 	associate
 	quotamod
 	quotaget
+	watch
 }
 
 class dir
@@ -163,9 +172,6 @@ inherits file
 	reparent
 	search
 	rmdir
-	open
-	audit_access
-	execmod
 }
 
 class file
@@ -173,52 +179,26 @@ inherits file
 {
 	execute_no_trans
 	entrypoint
-	execmod
-	open
-	audit_access
 }
 
 class lnk_file
 inherits file
-{
-	open
-	audit_access
-	execmod
-}
 
 class chr_file
 inherits file
 {
 	execute_no_trans
 	entrypoint
-	execmod
-	open
-	audit_access
 }
 
 class blk_file
 inherits file
-{
-	open
-	audit_access
-	execmod
-}
 
 class sock_file
 inherits file
-{
-	open
-	audit_access
-	execmod
-}
 
 class fifo_file
 inherits file
-{
-	open
-	audit_access
-	execmod
-}
 
 class fd
 {
diff --git a/public/global_macros b/public/global_macros
index 1a1d593b3..2c87fde5e 100644
--- a/public/global_macros
+++ b/public/global_macros
@@ -22,7 +22,7 @@ define(`ipc_class_set', `{ sem msgq shm ipc }')
 # Common groupings of permissions.
 #
 define(`x_file_perms', `{ getattr execute execute_no_trans map }')
-define(`r_file_perms', `{ getattr open read ioctl lock map }')
+define(`r_file_perms', `{ getattr open read ioctl lock map watch watch_reads }')
 define(`w_file_perms', `{ open append write lock map }')
 define(`rx_file_perms', `{ r_file_perms x_file_perms }')
 define(`ra_file_perms', `{ r_file_perms append }')
@@ -30,7 +30,7 @@ define(`rw_file_perms', `{ r_file_perms w_file_perms }')
 define(`rwx_file_perms', `{ rw_file_perms x_file_perms }')
 define(`create_file_perms', `{ create rename setattr unlink rw_file_perms }')
 
-define(`r_dir_perms', `{ open getattr read search ioctl lock }')
+define(`r_dir_perms', `{ open getattr read search ioctl lock watch watch_reads }')
 define(`w_dir_perms', `{ open search write add_name remove_name lock }')
 define(`ra_dir_perms', `{ r_dir_perms add_name write }')
 define(`rw_dir_perms', `{ r_dir_perms w_dir_perms }')
diff --git a/public/neverallow_macros b/public/neverallow_macros
index e2b6ed1af..59fa441d2 100644
--- a/public/neverallow_macros
+++ b/public/neverallow_macros
@@ -1,7 +1,7 @@
 #
 # Common neverallow permissions
 define(`no_w_file_perms', `{ append create link unlink relabelfrom rename setattr write }')
-define(`no_rw_file_perms', `{ no_w_file_perms open read ioctl lock }')
+define(`no_rw_file_perms', `{ no_w_file_perms open read ioctl lock watch watch_mount watch_sb watch_with_perm watch_reads }')
 define(`no_x_file_perms', `{ execute execute_no_trans }')
 define(`no_w_dir_perms',  `{ add_name create link relabelfrom remove_name rename reparent rmdir setattr write }')
 
-- 
2.17.1

